### PR TITLE
Remove `os.chdir()` method; add support for other key_file types other than RSA

### PR DESCRIPTION
**Pull request by Cloudac7**
For what has been mentioned in the WeChat Group, it is in need to support not only the RSA type key and thus it is add in this PR.

Also, I found unexpected `cwd` change in a dpdispatcher submission, which could be OK when tasks submitted in serial, whose `cwd` is just the same for each single task. However, it might be a severe bug when tasks submitted in a parallel way, which might result in single task that just start in a **WRONG** `work_path` and thus it could not find the files in the right place. For some workflow management system (such as [prefect](https://docs.prefect.io/)), the bug was observed.

While in the former implementation, `os.chdir()` is only used in 3 situations. For `os.symlink()`, it seems `dpdispatcher` just not using relevant path so there is no need to change `cwd`. For `tarfile.add()`, `abspath` could be used to find the file while using `arcname` parameter could recover the relevant path in the tar file.  For `subprocess.Popen()`, its `cwd` parameter is just the key to the door. So with just `os.path` dealing with the relevant path in the code, we could just drop `os.chdir()` out to make `cwd` not change in a python run.

The changed way is test to be OK in the Unittests and cluster environment for `LocalContext` and `SSHContext`. However, I am not sure whether it would be OK for `Bohrium` and more tests might be in need.

### CODE REVIEW

File: dpdispatcher/ssh_context.py [hunk 1]
```diff
@@ -135,17 +135,48 @@ def _setup_ssh(self):
...
+        else:
+            for keytype, name in [
+                (paramiko.RSAKey, "rsa"),
+                (paramiko.DSSKey, "dsa"),
+                (paramiko.ECDSAKey, "ecdsa"),
+                (paramiko.Ed25519Key, "ed25519"),
+            ]:
+                for directory in [".ssh", "ssh"]:
+                    full_path = os.path.expanduser(
+                        "~/{}/id_{}".format(directory, name)
```
Comment (njzjz) [inline dpdispatcher/ssh_context.py:167]: This breaks the compatibility on the Windows platform. Use os.path.join or Pathlib instead.

**Review by njzjz — CHANGES_REQUESTED** [review]

File: dpdispatcher/ssh_context.py (hunk 1 above)
Comment (Cloudac7) [reply dpdispatcher/ssh_context.py:167]: `os.path.expanduser` is just a Windows compatible implementation while add `ssh` in loop is an official implementation from `paramiko`.

Comment (njzjz) [reply dpdispatcher/ssh_context.py:167]: This is my result on Windows:
```py
PS C:\Program Files\Microsoft Visual Studio\2022\Community> python.exe
Python 3.10.8 (tags/v3.10.8:aaaf517, Oct 11 2022, 16:50:30) [MSC v.1933 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> import os
>>> os.path.expanduser("~/.ssh/id_rsa")
'C:\\Users\\njzjz/.ssh/id_rsa'
```

The correct output should be `'C:\\Users\\njzjz\\.ssh\\id_rsa'`.

Comment (Cloudac7) [reply dpdispatcher/ssh_context.py:167]: OK, I understand. Using `pathlib` might be a better choice.

**Review by njzjz — APPROVED** [review]

**Pull request merged**